### ISSUE TITLE
DefaultRecordUnmapper should cache AbstractRow and record type

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13647]
A recent benchmark has shown that a non-negligible amount of time is spent in `DefaultRecordMapper` to allocate new `AbstractRow` instances as well as look up `Class` instances for the `Tools::recordType`

![image](https://user-images.githubusercontent.com/734593/172594641-c7e7496f-055a-4f50-aee3-5dea3ff2359b.png)

jOOQ internals should never use `DSLContext::newRecord` to create records, but always something more specific that doesn't unwrap and re-wrap any `AbstractRow`

The original benchmark is from here:
https://github.com/jOOQ/jOOQ/issues/13646

**Comment by lukaseder** [comment]
Improving this changes a benchmark from:

```
Benchmark                                     Mode  Cnt      Score     Error  Units
MultisetVsJoinBenchmark.testMultisetAgg      thrpt    7  20959.015 ± 680.543  ops/sbetter
```

To

```
Benchmark                                 Mode  Cnt      Score      Error  Units
MultisetVsJoinBenchmark.testMultisetAgg  thrpt    7  21527.731 ± 1222.794  ops/s
```

I.e. roughly a 2-3% improvement on *every* (?) `DefaultRecordUnmapping` operation. Definitely worth doing.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.7 (https://github.com/jOOQ/jOOQ/issues/13648), and 3.15.11 (https://github.com/jOOQ/jOOQ/issues/13649)
